File: csharp/06-nested-if-hell/test/PackageValidatorTest.cs
using Xunit;

public class PackageValidatorTest
{
    [Fact]
    public void AllowsNormalPackage()
    {
        var validator = new PackageValidator();
        var result = validator.can_deliver(new Package { Weight = 10, Hazardous = false, Weekend = false });
        Assert.True(result.Allowed);
        Assert.Null(result.Warning);
    }

    [Fact]
    public void RejectsOverweight()
    {
        var validator = new PackageValidator();
        var result = validator.can_deliver(new Package { Weight = 60, Hazardous = false, Weekend = false });
        Assert.False(result.Allowed);
        Assert.Equal("Weight exceeded", result.Warning);
    }

    [Fact]
    public void RejectsHazardous()
    {
        var validator = new PackageValidator();
        var result = validator.can_deliver(new Package { Weight = 10, Hazardous = true, Weekend = false });
        Assert.False(result.Allowed);
        Assert.Equal("Hazardous material", result.Warning);
    }

    [Fact]
    public void RejectsWeekend()
    {
        var validator = new PackageValidator();
        var result = validator.can_deliver(new Package { Weight = 10, Hazardous = false, Weekend = true });
        Assert.False(result.Allowed);
        Assert.Equal("No weekend delivery", result.Warning);
    }

    [Fact]
    public void RejectsExtremeTemperature()
    {
        var validator = new PackageValidator();
        var result = validator.can_deliver(new Package { Weight = 10, Hazardous = false, Weekend = false, TemperatureRequired = 50 });
        Assert.False(result.Allowed);
        Assert.Equal("Temperature out of range", result.Warning);
    }

    [Fact]
    public void AllowsValidTemperature()
    {
        var validator = new PackageValidator();
        var result = validator.can_deliver(new Package { Weight = 10, Hazardous = false, Weekend = false, TemperatureRequired = 20 });
        Assert.True(result.Allowed);
    }

    [Fact]
    public void AllowsRemoteSmallPackage()
    {
        var validator = new PackageValidator();
        var result = validator.can_deliver(new Package { Weight = 15, Hazardous = false, Weekend = false, RemoteArea = true });
        Assert.True(result.Allowed);
        Assert.Equal("Remote surcharge applies", result.Warning);
    }

    [Fact]
    public void RejectsRemoteHeavyPackage()
    {
        var validator = new PackageValidator();
        var result = validator.can_deliver(new Package { Weight = 25, Hazardous = false, Weekend = false, RemoteArea = true });
        Assert.False(result.Allowed);
        Assert.Equal("Too heavy for remote", result.Warning);
    }

    [Fact]
    public void RejectsNullPackage()
    {
        var validator = new PackageValidator();
        var result = validator.can_deliver(null);
        Assert.False(result.Allowed);
        Assert.Equal("No package", result.Warning);
    }

    [Fact]
    public void RejectsMissingWeight()
    {
        var validator = new PackageValidator();
        var result = validator.can_deliver(new Package { Hazardous = false, Weekend = false });
        Assert.False(result.Allowed);
        Assert.Equal("No weight specified", result.Warning);
    }

    [Fact]
    public void Weight50AllowedAtBoundary()
    {
        var validator = new PackageValidator();
        var result = validator.can_deliver(new Package { Weight = 50, Hazardous = false, Weekend = false });
        Assert.True(result.Allowed);
        Assert.Null(result.Warning);
    }

    [Fact]
    public void Temperature40AllowedAtBoundary()
    {
        var validator = new PackageValidator();
        var result = validator.can_deliver(new Package { Weight = 10, Hazardous = false, Weekend = false, TemperatureRequired = 40 });
        Assert.True(result.Allowed);
    }

    [Fact]
    public void TemperatureMinus20AllowedAtBoundary()
    {
        var validator = new PackageValidator();
        var result = validator.can_deliver(new Package { Weight = 10, Hazardous = false, Weekend = false, TemperatureRequired = -20 });
        Assert.True(result.Allowed);
    }

    [Fact]
    public void RemoteWeight20AllowedAtBoundary()
    {
        var validator = new PackageValidator();
        var result = validator.can_deliver(new Package { Weight = 20, Hazardous = false, Weekend = false, RemoteArea = true });
        Assert.True(result.Allowed);
        Assert.Equal("Remote surcharge applies", result.Warning);
    }
}

File: python/06-nested-if-hell/test/package_validator_test.py
from src.package_validator import PackageValidator

def test_normal_package():
    validator = PackageValidator()
    result = validator.can_deliver({"weight": 10, "hazardous": False, "weekend": False})
    assert result["allowed"] is True
    assert result["warning"] is None

def test_overweight():
    validator = PackageValidator()
    result = validator.can_deliver({"weight": 60, "hazardous": False, "weekend": False})
    assert result["allowed"] is False
    assert result["warning"] == "Weight exceeded"

def test_hazardous():
    validator = PackageValidator()
    result = validator.can_deliver({"weight": 10, "hazardous": True, "weekend": False})
    assert result["allowed"] is False
    assert result["warning"] == "Hazardous material"

def test_weekend():
    validator = PackageValidator()
    result = validator.can_deliver({"weight": 10, "hazardous": False, "weekend": True})
    assert result["allowed"] is False
    assert result["warning"] == "No weekend delivery"

def test_extreme_temperature():
    validator = PackageValidator()
    result = validator.can_deliver({"weight": 10, "hazardous": False, "weekend": False, "temperatureRequired": 50})
    assert result["allowed"] is False
    assert result["warning"] == "Temperature out of range"

def test_valid_temperature():
    validator = PackageValidator()
    result = validator.can_deliver({"weight": 10, "hazardous": False, "weekend": False, "temperatureRequired": 20})
    assert result["allowed"] is True

def test_remote_small():
    validator = PackageValidator()
    result = validator.can_deliver({"weight": 15, "hazardous": False, "weekend": False, "remoteArea": True})
    assert result["allowed"] is True
    assert result["warning"] == "Remote surcharge applies"

def test_remote_heavy():
    validator = PackageValidator()
    result = validator.can_deliver({"weight": 25, "hazardous": False, "weekend": False, "remoteArea": True})
    assert result["allowed"] is False
    assert result["warning"] == "Too heavy for remote"

def test_null_package():
    validator = PackageValidator()
    result = validator.can_deliver(None)
    assert result["allowed"] is False
    assert result["warning"] == "No package"

def test_missing_weight():
    validator = PackageValidator()
    result = validator.can_deliver({"hazardous": False})
    assert result["allowed"] is False
    assert result["warning"] == "No weight specified"


def test_weight_50_allowed():
    validator = PackageValidator()
    result = validator.can_deliver({"weight": 50, "hazardous": False, "weekend": False})
    assert result["allowed"] is True
    assert result["warning"] is None


def test_temperature_40_allowed():
    validator = PackageValidator()
    result = validator.can_deliver({"weight": 10, "hazardous": False, "weekend": False, "temperatureRequired": 40})
    assert result["allowed"] is True


def test_temperature_minus_20_allowed():
    validator = PackageValidator()
    result = validator.can_deliver({"weight": 10, "hazardous": False, "weekend": False, "temperatureRequired": -20})
    assert result["allowed"] is True


def test_remote_weight_20_allowed():
    validator = PackageValidator()
    result = validator.can_deliver({"weight": 20, "hazardous": False, "weekend": False, "remoteArea": True})
    assert result["allowed"] is True
    assert result["warning"] == "Remote surcharge applies"
